File: #1.data-structures/#2.Stack/src/java/src/modules/UndoRedo.java
public class UndoRedo<T> {
  private final Stack<T> undoStack = new Stack<>();
  private Stack<T> redoStack = new Stack<>();

  public void addAction(T action) {
      undoStack.push(action);
      redoStack = new Stack<T>(); // Limpa a pilha de redo
  }

  public T undo() {
      T action = undoStack.pop();
      if (action != null) {
          redoStack.push(action);
      }
      return action;
  }

  public T redo() {
      T action = redoStack.pop();
      if (action != null) {
          undoStack.push(action);
      }
      return action;
  }

  public void debug() {
      System.out.println("Undo Stack: " + undoStack.getItems());
      System.out.println("Redo Stack: " + redoStack.getItems());
  }
}

File: #1.data-structures/#2.Stack/src/python/src/modules/UndoRedo.py
from typing import Generic, TypeVar, Optional
from modules.Stack import Stack  # Assumindo que o Stack está implementado em um arquivo separado chamado stack.py

T = TypeVar('T')

class UndoRedo(Generic[T]):
    def __init__(self):
        self._undo_stack: Stack[T] = Stack[T]()
        self._redo_stack: Stack[T] = Stack[T]()

    def add_action(self, action: T) -> None:
        self._undo_stack.push(action)
        self._redo_stack = Stack()  # Limpa a pilha de redo

    def undo(self) -> Optional[T]:
        action = self._undo_stack.pop()
        if action is not None:
            self._redo_stack.push(action)
        return action

    def redo(self) -> Optional[T]:
        action = self._redo_stack.pop()
        if action is not None:
            self._undo_stack.push(action)
        return action

    def debug(self) -> None:
        print("Undo Stack:", list(self._undo_stack.items))
        print("Redo Stack:", list(self._redo_stack.items))
